<a href="https://colab.research.google.com/github/HolaBaGa/tibame/blob/main/%E5%AE%8C%E5%85%A8%E5%B9%B3%E6%96%B9%E6%97%A5_0429.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# 定義函數何謂閏年
def is_leap_year(year):
    if year % 400 == 0:
        return True
    elif year % 100 == 0:
        return False
    elif year % 4 == 0:
        return True
    else:
        return False

In [ ]:
# 定義函數判斷正確的年月日
# 20000831(O) 20001331(X) 20000229(O) 20010229(X)
# is_date_valid()
def is_date_valid(d):
    day = d % 100
    remain = d // 100
    month = remain % 100
    year = remain // 100
    if year >= 1:
        if 12 >= month >= 1:
            month_days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
            max_days = month_days[month-1]
            # if is_leap_year(year) == True
            if is_leap_year(year) and month == 2:
                max_days = max_days + 1
            if max_days >= day >= 1:
                return True
            else:
                return False
        else:
            return False
    else:
        return False
print(is_date_valid(20000229))
print(is_date_valid(20010229))


True
False


In [ ]:
# 想法2 最快速能找到答案，運用+1平方的方式
# 20241001 2072...
start = 4499
result = []
while True:
    d = start * start
    if is_date_valid(d):
        # list專屬功能: append
        # l = l.append(X) l.append(O)
        result.append(d)
    if len(result) == 10:
        break
print(result)
start = start + 1

[20241001, 20241001, 20241001, 20241001, 20241001, 20241001, 20241001, 20241001, 20241001, 20241001]


走過一群東西(list/str): for 名稱 in 群

固定次數: for i in range(次)

其他條件: while True + break

In [ ]:
# 想法2 最快速能找到答案
# 20241001 2072...
def fully_squared_2(start=4499, max_count=10):
    result = []
    while True:
        d = start * start
        if is_date_valid(d):
            # list專屬功能: append
            # l = l.append(X) l.append(O)
            result.append(d)
        if len(result) == max_count:
            break
        start = start + 1
    return result
print(fully_squared_2(max_count=5))

[20241001, 20720704, 20930625, 21150801, 21390625]


20241001 -> 20241002 -> 20241003

python內建處理 a. datetime: 日期與時間 b. time: 時間

In [ ]:
#資料科學常使用
# 型態轉換: str - 日期型態
# strptime: str -> 日期
# strftime: 日期 -> str
import datetime as dt
s = "2000-02-29"
d = dt.datetime.strptime(s, "%Y-%m-%d")
s2 = d.strftime("紀念日: %Y/%m/%d(%a)")
print(s2)

紀念日: 2000/02/29(Tue)


In [ ]:
# 如何準備一個日期
d = dt.datetime(2000, 2, 28)
# 增加/減少: timedelta
delta = dt.timedelta(days=1, minutes=20)
d + 2 * delta


datetime.datetime(2000, 3, 1, 0, 40)

問題在於: 你如何確定一個小數接近一個整數

錯誤: 你覺得3.0才跟3一樣, 正確 3.0000001其實就跟3.0一樣道理

小數是一個不精確值: 兩個小數只要差不多, 他們就是相等

一塊pizza(1.0)分三塊(0.3333333) -> 0,33333333 * 3 = 0.999999999 ?= 1.0

3.00000000000049 == 3.00000000000000005

如果自己寫: abs(小數1-小數2) <= 某個很小的值(e.g. 10**-7)

In [ ]:
a = 3.0000000000000001
a.is_integer()


True

In [ ]:
b = 3+10**-16
b.is_integer()

False

In [ ]:
# 想法1.從日期慢慢+1
import datetime as dt
d = dt.datetime(2024, 10, 1)
delta = dt.timedelta(days=1)
result = []
# 迴圈
while True:
    # 把日期往後一個一個加
    # 日期->字串->整數
    d_int = int(d.strftime("%Y%m%d"))
    # 開根號看是否為整數
    if (d_int ** 0.5).is_integer():
        # 把完全平方日日期轉成的整數放到result
        result.append(d_int)
    if len(result) == 10:
        break
    d = d + delta
print(result)


[20241001, 20720704, 20930625, 21150801, 21390625, 21520321, 21641104, 21650409, 21790224, 21921124]


In [ ]:
def fully_squared_1(year=2024, month=10, day=1, max_count=10):
    d = dt.datetime(year, month, day)
    delta = dt.timedelta(days=1)
    result = []
    # 迴圈
    while True:
        # 把日期往後一個一個加
        # 日期->字串->整數
        d_int = int(d.strftime("%Y%m%d"))
        # 開根號看是否為整數
        if (d_int ** 0.5).is_integer():
            # 把完全平方日日期轉成的整數放到result
            result.append(d_int)
        if len(result) == max_count:
            break
        d = d + delta
    return result

fully_squared_1(max_count=20)

[20241001,
 20720704,
 20930625,
 21150801,
 21390625,
 21520321,
 21641104,
 21650409,
 21790224,
 21921124,
 22071204,
 22080601,
 22231225,
 22420225,
 22591009,
 22600516,
 22800625,
 23011209,
 23020804,
 23030401]

In [ ]:
# 在電腦裡算時間是使用秒數 Unix timestamp  使用格林威治時間
import time

total = 300

t1 = time.time()
fully_squared_1(max_count=total)
t2 = time.time()
dif1 = t2 - t1
print("想法1:", dif1, "秒")

t1 = time.time()
fully_squared_2(max_count=total)
t2 = time.time()
dif2 = t2 - t1
print("想法2:", dif2, "秒")

ratio = dif1 / dif2
print("相差", ratio, "倍")